### PR TITLE
chore: allow pass row fd to capture

### PR DESCRIPTION
**Pull request by Decodetalkers**
https://github.com/pop-os/xdg-desktop-portal-cosmic/blob/master_jammy/src/wayland.rs#L191

from the code in xdg-desktop-portal-cosmis, here should pass a rawfd to it , so maybe expose a function is better?

### CODE REVIEW

File: libwayshot/src/lib.rs [hunk 1]
```diff
@@ -142,6 +150,147 @@ impl WayshotConnection {
...
+        // Instantiating screencopy manager.
+        let screencopy_manager = match self.globals.bind::<ZwlrScreencopyManagerV1, _, _>(
+            &qh,
+            3..=3,
+            (),
+        ) {
+            Ok(x) => x,
+            Err(e) => {
+                log::error!("Failed to create screencopy manager. Does your compositor implement ZwlrScreencopy?");
+                panic!("{:#?}", e);
```
Comment (Decodetalkers) [inline libwayshot/src/lib.rs:180]: e.. I think here should not panic.. if is used in xdg-desktop-portal

**Comment by Decodetalkers** [comment]
let me test it weather it will work

File: libwayshot/src/lib.rs (hunk 1 above)
Comment (Shinyzenith) [reply libwayshot/src/lib.rs]: That is a good point, we can return an error here?

**Comment by Decodetalkers** [comment]
https://github.com/waycrate/wayshot/assets/60290287/a1b81cc8-aea9-4401-8f0e-78733a95fed6
Finshed, so when pass raw fd, it should not use memfd

**Comment by Decodetalkers** [comment]
I think this need https://github.com/waycrate/wayshot/pull/59 , we need to bind wloutput first, and wloutput dispatch too much time

File: libwayshot/src/error.rs [hunk 2]
```diff
@@ -21,4 +21,10 @@ pub enum Error {
     Global(#[from] GlobalError),
     #[error("connect error: {0}")]
     Connect(#[from] ConnectError),
+    #[error("framecopy failed")]
+    FramecopyFailed,
+    #[error("No supported buffer format")]
+    NoSupportedBufferFormat,
+    #[error("Cannot find Wayland protocols")]
```
Comment (Shinyzenith) [inline libwayshot/src/error.rs:28]:
```suggestion
    #[error("Cannot find required wayland protocol")]
```

File: libwayshot/src/lib.rs [hunk 3]
```diff
@@ -159,13 +164,26 @@ impl WayshotConnection {
     }
 
     /// Get a FrameCopy instance with screenshot pixel data for any wl_output object.
-    fn capture_output_frame(
+    /// And write data to fd
+    /// It can be used in screencast
```
Comment (Shinyzenith) [inline libwayshot/src/lib.rs:168]:
```suggestion
```

File: libwayshot/src/lib.rs [hunk 4]
```diff
@@ -159,13 +164,26 @@ impl WayshotConnection {
     }
 
     /// Get a FrameCopy instance with screenshot pixel data for any wl_output object.
-    fn capture_output_frame(
+    /// And write data to fd
```
Comment (Shinyzenith) [inline libwayshot/src/lib.rs:167]:
```suggestion
    ///  Data will be written to fd.
```

**Review by Shinyzenith — CHANGES_REQUESTED** [review]

**Review by Shinyzenith — APPROVED** [review]

**Pull request merged**